## 1.  What is a List?

A list is a mutable heterogenous ordered data structure in python that allows duplicates. For example - 

In [25]:
my_list = [1,2,3,4,5,4]
heterogeneous_list = [1, "two", 3.0, True, None]
my_list.append(6) #mutable
print(my_list)
print(heterogeneous_list)
print(my_list[0:-2])


[1, 2, 3, 4, 5, 4, 6]
[1, 'two', 3.0, True, None]
[1, 2, 3, 4, 5]


## 2.  What is a Tuple? 

A tuple is an ordered, heterogenous, immutable data structure in python that allows duplicates. For example - 

In [26]:
my_tuple = (1,2,3,4,5,6)
heterogeneous_tuple = (1,"two", 3.0, True, None)
my_tuple = my_tuple + (7,8) #immutable
print(my_tuple)
print(heterogeneous_tuple)


(1, 2, 3, 4, 5, 6, 7, 8)
(1, 'two', 3.0, True, None)


## 3.  What is the difference between List and Tuple? 

List is a mutable data structure while tuple is an immutable data structure. This means that you can change the contents of a list after it has been created, while you cannot change the contents of a tuple after it has been created.
List is denoted by []
Tuple is denoted by ()

In [27]:
my_list = [1,2,3,4,5]
my_tuple = (1,2,3,4,5)

my_list[0] = 10 #mutable
print(my_list)
my_tuple[0] = 10 #immutable, error will be raised
print(my_tuple)

[10, 2, 3, 4, 5]


TypeError: 'tuple' object does not support item assignment

## 4.  Python Program to find the largest element in the list.

In [ ]:
my_list = [1,2,3,4,5]
max_elem = max(my_list)
print(f"The maximum element is {max_elem}")



The maximum element is 5


## 5.  Python program to interchange first and last elements in a list. 

In [ ]:
my_list = [1,2,3,4,5]
temp = my_list[-1];
my_list[-1] = my_list[0];
my_list[0] = temp;
print(f"The list after the interchange is :  {my_list}")

The list after the interchange is :  [5, 2, 3, 4, 1]


In [ ]:
my_list = [1,2,3,4,5]
my_list[0], my_list[-1] = my_list[-1], my_list[0] ## swapping using tuple unpacking in python
print(f"The list after interchange is {my_list}")

The list after interchange is [5, 2, 3, 4, 1]


## 6. Python program to swap two elements in a list

In [36]:
my_list = [1,2,3,4,5]
def swapTwoElements(my_list,i,j):
    my_list[i], my_list[j] = my_list[j], my_list[i]

swapTwoElements(my_list,0,-1)
print(f"The list after interchange is : {my_list}")
    



The list after interchange is : [5, 2, 3, 4, 1]


## 7. Python program to Reverse a List

In [37]:
my_list = [12,10,120,1,2]
my_list.reverse()
print(f"The reveresed list is : {my_list}")

The reveresed list is : [2, 1, 120, 10, 12]


## 8. Python program to count occurrences of an element in a list

In [38]:
my_list = [1,2,3,3,4,4,5,5,5,6]
def countOccurenceOfAnElement(my_list,elem):
    return my_list.count(elem);

print(f"The count of 3 in {my_list} is {countOccurenceOfAnElement(my_list=my_list,elem=3)}")

The count of 3 in [1, 2, 3, 3, 4, 4, 5, 5, 5, 6] is 2


## 9. Python program to find the sum of elements in a list

In [39]:
my_list = [1,2,3,4,5]
print(f"The sum of the elements in the list {my_list} is {sum(my_list)}")

The sum of the elements in the list [1, 2, 3, 4, 5] is 15


## 10. Python program to Multiply all numbers in the list

In [45]:
my_list = [1,2,3,4,5]
productOfList = 1
for num in my_list:
    productOfList = productOfList*num

print(f"The product of all the elements in the list {my_list} is {productOfList}")

## Another way to do is - 
from math import prod
print(f"The product of all the elements in the list {my_list} is {prod(my_list)}")

The product of all the elements in the list [1, 2, 3, 4, 5] is 120
The product of all the elements in the list [1, 2, 3, 4, 5] is 120


## 11. What are the ways to find the length of a list

In [ ]:
my_list = [1,2,3,4,5]

print(f"The len of the list {my_list} is {len(my_list)}")


The len of the list [1, 2, 3, 4, 5] is 5


## 12. Python program to find the smallest and largest number in a list (Without min-max function)

In [49]:
import math

my_list = [1,2,3,4,5,6]

def maxElem(my_list):
    max_elem = -math.inf
    for num in my_list:
        if num > max_elem:
            max_elem = num

    return max_elem

def minElem(my_list):
    min_elem = math.inf
    for num in my_list:
        if num < min_elem:
            min_elem = num

    return min_elem     

print(f"The max element in the list {my_list} is {maxElem(my_list=my_list)}")  
print(f"The min element in the list {my_list} is {minElem(my_list=my_list)}")  


The max element in the list [1, 2, 3, 4, 5, 6] is 6
The min element in the list [1, 2, 3, 4, 5, 6] is 1


## 13. Python Program to find the area of a circle

In [ ]:
def calculateAreaOfCircle(radius):
    return math.pi * radius * radius

print(f"The area of a circle having radius of 3cm is {round(calculateAreaOfCircle(3),2)} sqcm")

The area of a circle having radius of 3cm is 28.27 sqcm


## 14. Take inputs from the user to make a list. 
## Again take one input from the user and delete that  element from a list.

In [ ]:
my_list = list(map(int, input("Enter the numbers separated by comma -").split(",")))

print(f"Your entered list is {my_list}")

elemToDelete = int(input("Enter an elem you want to delete from the list - "))

my_list.remove(elemToDelete)

print(f"The list after deletion of {elemToDelete} is {my_list}");

## Note the remove function will remove the first occurence of the element from the list.

Your entered list is [1, 2, 4, 5, 6]
The list after deletion of 1 is [2, 4, 5, 6]


## 15. You are given a list of integer elements. Make a new list that will store a square of elements
## of the previous list. (With and without list comprehension)
## ● Input_list = [2,5,6,12]
## ● Output_list = [4,25,36,144]

In [60]:
input_list = [2,5,6,12]

## Without list comprehension
output_list = []

for num in input_list:
    output_list.append(num*num)

print(f"The output list without list comprehension is {output_list}")


## With list comprehension 

output_list_lc = [elem*elem for elem in input_list]
print(f"The output list with list comprehension is {output_list_lc}");


The output list without list comprehension is [4, 25, 36, 144]
The output list with list comprehension is [4, 25, 36, 144]


## 16. WAP to create two lists, one containing all even numbers 
## and the other containing all odd numbers between 0 to 151 

In [63]:
even_list = [elem for elem in range(0,152) if elem % 2 == 0]
odd_list = [elem for elem in range(0,152) if elem % 2 != 0]

print(f"Even number list is {even_list}")
print(f"Odd number list is {odd_list}")

Even number list is [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150]
Odd number list is [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151]


## 17. Python program to Count Even and Odd numbers in a List

In [65]:
my_list = [1,2,3,4,5,6,7,8,9,10,11]

even_count = 0;
odd_count = 0;

for num in my_list:
    if(num % 2 == 1):
        odd_count+=1
    else:
        even_count+=1 

print(f"The even count for list {my_list} is {even_count}")        
print(f"The odd count for list {my_list} is {odd_count}")        

The even count for list [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11] is 5
The odd count for list [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11] is 6


## 18. WAP to make new lists, containing only numbers 
## which are divisible by 4, 6, 8, 10, 3, 5, 7, and 9 
## in separate lists for range(0,151) 

In [76]:
## If we need one list containing all elements.
my_list = [elem for elem in range(0,152) if (elem % 4 == 0 or elem % 6 == 0 or elem % 8 == 0 or elem % 10 == 0 or  elem % 3 == 0 or elem % 5 == 0 or elem % 7 == 0 or elem % 9 == 0)]
print(f"The list is {my_list}")

## If we need separate lists for separate divisors we can do as below - 
divisors = [4,6,8,10,3,5,7,9]
result_lists = []

for i in range(0,len(divisors)):
    divisor = divisors[i]
    divisor_list = [elem for elem in range(0,152) if elem % divisor == 0]
    result_lists.append(divisor_list)


print(f"The results lists in the same order is {result_lists}");


The list is [0, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 24, 25, 27, 28, 30, 32, 33, 35, 36, 39, 40, 42, 44, 45, 48, 49, 50, 51, 52, 54, 55, 56, 57, 60, 63, 64, 65, 66, 68, 69, 70, 72, 75, 76, 77, 78, 80, 81, 84, 85, 87, 88, 90, 91, 92, 93, 95, 96, 98, 99, 100, 102, 104, 105, 108, 110, 111, 112, 114, 115, 116, 117, 119, 120, 123, 124, 125, 126, 128, 129, 130, 132, 133, 135, 136, 138, 140, 141, 144, 145, 147, 148, 150]
The results lists in the same order is [[0, 4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 144, 148], [0, 6, 12, 18, 24, 30, 36, 42, 48, 54, 60, 66, 72, 78, 84, 90, 96, 102, 108, 114, 120, 126, 132, 138, 144, 150], [0, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96, 104, 112, 120, 128, 136, 144], [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150], [0, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72,

## 19.  Given a list containing elements of different types (integers, strings, and floats),
## create three separate lists to store integers, strings, and floats respectively. 


In [ ]:
my_list = [1,"vijay", 1.34, 2,"john", 3.34, 11,"jane", 1.43]
integer_list = [elem for elem in my_list if type(elem) == int]
float_list = [elem for elem in my_list if type(elem) == float]
string_list = [elem for elem in my_list if type(elem) == str]

print(integer_list, float_list, string_list)

## We can use isinstance instead of type as it is more pythonic and handles subclasses.

# my_list = [1, "vijay", 1.34, 2, "john", 3.34, 11, "jane", 1.43]

# integer_list = [x for x in my_list if isinstance(x, int)]
# float_list   = [x for x in my_list if isinstance(x, float)]
# string_list  = [x for x in my_list if isinstance(x, str)]

# print(integer_list, float_list, string_list)


[1, 2, 11] [1.34, 3.34, 1.43] ['vijay', 'john', 'jane']


## 20.  What’s The Difference Between The Python append() and extend() Methods?

In [10]:
my_list = [1,2,4,4,5,6]
my_list.append([10,11,12]) ## append adds one object as a single item at the end of a list.
print(my_list)
my_list.extend([100,99,98]) ## extend adds each item from an iterable into the list.
print(my_list)

[1, 2, 4, 4, 5, 6, [10, 11, 12]]
[1, 2, 4, 4, 5, 6, [10, 11, 12], 100, 99, 98]


## 21.  Write a Python program to append a list to the second list 


In [11]:
list_to_be_appended = [1,5,7,9,11]
list_to_append = [2,4,6,7,8]
list_to_append.append(list_to_be_appended)
print(list_to_append)

[2, 4, 6, 7, 8, [1, 5, 7, 9, 11]]


## 22.  Write a Python program to find the third-largest number in a list 


In [ ]:
my_list = [1,3,4,5,2,9]
sorted_list = sorted(my_list, reverse= True)
print(f"The third largest number in the list is {sorted_list[2]}")    

The third largest number in the list is 4


## 23.  Write a Python program to check whether a list contains a sublist 

In [21]:
my_list = [1,2,4,5,[5,6,7,8]]
ans = False
for x in my_list:
    if(isinstance(x,list)):
        ans = True
        break
if(ans == True):
    print("The list contains a sublist.")
else:
    print("The list does not contains a sublist.")


The list contains a sublist.


## 24.  Write a Python program to generate all sublists of a list 

In [36]:
my_list = [1,2,4,5,6,7,8]


# for i in range(0, len(my_list)):
#     for j in range(i, len(my_list)):
#         sub_list = []
#         for k in range(i,j+1):
#             sub_list.append(my_list[k])
#         print(sub_list)    

## In python we  can also do -  
for i in range(0, len(my_list)):
    for j in range(i, len(my_list)):
        print(my_list[i:j+1])



[1]
[1, 2]
[1, 2, 4]
[1, 2, 4, 5]
[1, 2, 4, 5, 6]
[1, 2, 4, 5, 6, 7]
[1, 2, 4, 5, 6, 7, 8]
[2]
[2, 4]
[2, 4, 5]
[2, 4, 5, 6]
[2, 4, 5, 6, 7]
[2, 4, 5, 6, 7, 8]
[4]
[4, 5]
[4, 5, 6]
[4, 5, 6, 7]
[4, 5, 6, 7, 8]
[5]
[5, 6]
[5, 6, 7]
[5, 6, 7, 8]
[6]
[6, 7]
[6, 7, 8]
[7]
[7, 8]
[8]


## 25.  Write a Python program to find common items from two lists 

In [38]:
list1 = [1,2,3,4,5,6]
list2 = [98,99,3,4,50,60]

common_elements = [elem for elem in list1 if elem in list2]
print(common_elements)



[3, 4]


## 26. How to flatten a list in Python?

In [ ]:
my_list = [1,2,3,[4,5,6], [7,[8,9]]];

def flatten_list(input_list):
    flattened_list = []
    for elem in input_list:
        if isinstance(elem, list):
            flattened_list.extend(flatten_list(elem))
        else:
            flattened_list.append(elem)

    return flattened_list

print(f"The flattened list is {flatten_list(my_list)}")




The flattened list is [1, 2, 3, 4, 5, 6, 7, 8, 9]


## 27. How to sort a list in ascending and descending order without using the sort function?

In [6]:
my_list = [8,3,1,2,0,5,4, 9,7,6]

def bubble_sort(list_to_sort, ascending = True):
    n = len(list_to_sort)
    for i in range(n):
        for j in range(0, n-i-1):
            to_swap = list_to_sort[j] > list_to_sort[j + 1] if ascending == True else list_to_sort[j] < list_to_sort[j + 1]
            if(to_swap):
                list_to_sort[j], list_to_sort[j+1] = list_to_sort[j+1], list_to_sort[j]

bubble_sort(my_list,ascending = False)
print(f"The sorted list is descending order is - {my_list}")
bubble_sort(my_list,ascending = True)
print(f"The sorted list is ascending order is - {my_list}")




The sorted list is descending order is - [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
The sorted list is ascending order is - [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


## 28. How to sort a tuple?

In [12]:
my_tuple = (8,3,1,2,0,5,4, 9,7,6)
sorted_tuple = tuple(sorted(my_tuple))
print(f"The sorted tuple is {sorted_tuple}")

my_tuple.sort() ## This will give an error as tuple is immutable and does not have a sort method.


The sorted tuple is (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)


AttributeError: 'tuple' object has no attribute 'sort'

## 29. Write a Python program to convert a list of multiple integers into a single integer
## ● [11, 33, 50] >>> 113350

In [14]:
my_list = [11,33,50]

output = ""

for elem in my_list:
    output += str(elem)

print(output)

113350


## 30. What is the difference between del and clear?

In [17]:
my_list = [1,3,5,6,7,8,9]
del my_list[0] ## deletes only the first element 
print(my_list)
my_list.clear() ## removes all elements 
print(my_list)

[3, 5, 6, 7, 8, 9]
[]


## 31. What is the difference between remove and pop?

In [21]:
my_list  = [1,2,3,4,5]

my_list.pop() ## removes the last element of the list 

print(my_list)

my_list.remove(3) ## removes the element we specify 

print(my_list)

[1, 2, 3, 4]
[1, 2, 4]


## 32. What is the difference between indexing and Slicing?

In [23]:
my_list = [1,2,3,4,5,6,7,8,9]
print(my_list[0:4]) ## slicing gives us a sublist  
print(my_list[0]) ## indexing gives us one element 



[1, 2, 3, 4]
1


## 33. What is the difference between sort and sorted?

In [24]:
my_list = [4,23,12,1,2]
sorted_list = sorted(my_list) ## gives us new sorted list , does not mutate the original list 
print(sorted_list)
my_list.sort() ## sorts the list in place 
print(my_list)

[1, 2, 4, 12, 23]
[1, 2, 4, 12, 23]


## 34. What is the difference between reverse and reversed function?

In [ ]:
my_list = [1,2,3,4,5,6]

my_list.reverse() ## reverses the list in place

print(my_list)

print(list(reversed(my_list))) ## returns the reversed list as a new list

[6, 5, 4, 3, 2, 1]
[1, 2, 3, 4, 5, 6]


## 35. What is the difference between copy and deep copy?

In [34]:
from copy import deepcopy

my_list = [1,2,3,4,5,[6,7,8,9]]

copied_list  =  my_list.copy() ## returns a shallow copy of the list
deep_copied_list = deepcopy(my_list) ## returns a deep copy of the list
my_list[5].append(10)

print(copied_list) ## A shallow copy creates a new outer list, but nested objects are still shared, so when we make changes in nested list in my_list, it will also reflect in copied_list.
print(deep_copied_list) ## A deep copy creates a new outer list and recursively copies all nested objects, so changes in my_list will not affect deep_copied_list.


[1, 2, 3, 4, 5, [6, 7, 8, 9, 10]]
[1, 2, 3, 4, 5, [6, 7, 8, 9]]
